### PR TITLE
RetroAchievements - On Screen Images

### PR DESCRIPTION
**Pull request by LillyJadeKatrin**
This is a portion of an integration with the RetroAchievements tools and libraries for connecting to the website, downloading data, unlocking achievements and submitting to leaderboards for games running in Dolphin. In this PR, I add to the Badges functionality to incorporate badges in the OnScreenDisplay messages that pop up during various achievement-related events.

### CODE REVIEW

**Comment by LillyJadeKatrin** [comment]
Note that at this time, the first six commits in this PR are from https://github.com/dolphin-emu/dolphin/pull/12025.

**Comment by LillyJadeKatrin** [comment]
Rebased to remove the commits from #12025 now that it has been merged into master.

File: Source/Core/Core/AchievementManager.cpp [hunk 1]
```diff
@@ -1086,6 +1081,54 @@ AchievementManager::PingRichPresence(const RichPresence& rich_presence)
...
+    return badge_status->texture;
+  float width = BADGE_HW;
+  float height = BADGE_HW;
+  VideoCommon::CustomTextureData::ArraySlice::Level level;
+  if (!Common::LoadPNG(badge_status->badge, &level.data, &level.width, &level.height))
+  {
+    ERROR_LOG_FMT(ACHIEVEMENTS, "Failed to load texture for badge {}.", badge_status->name);
+    return nullptr;
+  }
+  level.format = AbstractTextureFormat::RGBA8;
```
Comment (iwubcode) [inline Source/Core/Core/AchievementManager.cpp:1122]: Apologies for the wrong direction, these two lines aren't really necessary after all (they don't hurt but might as well remove them)

File: Source/Core/VideoCommon/OnScreenDisplay.cpp [hunk 2]
```diff
@@ -48,8 +50,8 @@ struct Message
   bool ever_drawn = false;
   bool should_discard = false;
   u32 color = 0;
-  std::unique_ptr<Icon> icon;
-  std::unique_ptr<AbstractTexture> texture;
+  std::shared_ptr<Icon> icon;
+  std::shared_ptr<AbstractTexture> texture;
```
Comment (AdmiralCurtiss) [inline Source/Core/VideoCommon/OnScreenDisplay.cpp:54]: Why did you change this, what exactly are you trying to accomplish by doing so? Same for all other changes in this file.

File: Source/Core/Core/AchievementManager.h [hunk 3]
```diff
@@ -188,6 +199,7 @@ class AchievementManager
   Common::WorkQueueThread<std::function<void()>> m_queue;
   Common::WorkQueueThread<std::function<void()>> m_image_queue;
   std::recursive_mutex m_lock;
+  std::recursive_mutex m_video_lock;
```
Comment (AdmiralCurtiss) [inline Source/Core/Core/AchievementManager.h:202]: There's no need for this lock as far as I can tell, all usages are already locked by `m_lock`.

File: Source/Core/Core/AchievementManager.h [hunk 4]
```diff
@@ -58,11 +63,13 @@ class AchievementManager
   static constexpr size_t RP_SIZE = 256;
   using RichPresence = std::array<char, RP_SIZE>;
   using Badge = std::vector<u8>;
+  static constexpr u32 BADGE_HW = 96;
```
Comment (AdmiralCurtiss) [inline Source/Core/Core/AchievementManager.h:66]: This is not used.

File: Source/Core/Core/AchievementManager.h [hunk 5]
```diff
@@ -19,6 +21,9 @@
 #include "Common/Event.h"
 #include "Common/WorkQueueThread.h"
 
+#include "VideoCommon/AbstractTexture.h"
```
Comment (AdmiralCurtiss) [inline Source/Core/Core/AchievementManager.h:24]: This is not used.

File: Source/Core/Core/AchievementManager.h [hunk 6]
```diff
@@ -119,6 +126,7 @@ class AchievementManager
   void AchievementEventHandler(const rc_runtime_event_t* runtime_event);
 
   std::recursive_mutex* GetLock();
+  std::recursive_mutex* GetVideoLock();
```
Comment (AdmiralCurtiss) [inline Source/Core/Core/AchievementManager.h:129]: This is not used and doesn't need to be exposed.

File: Source/Core/Core/AchievementManager.h [hunk 7]
```diff
@@ -178,6 +188,7 @@ class AchievementManager
   u32 m_game_id = 0;
   rc_api_fetch_game_data_response_t m_game_data{};
   bool m_is_game_loaded = false;
+  u32 m_framecount = false;
```
Comment (AdmiralCurtiss) [inline Source/Core/Core/AchievementManager.h:191]: You're assigning a boolean to an integer here.

File: Source/Core/Core/AchievementManager.h [hunk 8]
```diff
@@ -58,11 +63,13 @@ class AchievementManager
   static constexpr size_t RP_SIZE = 256;
   using RichPresence = std::array<char, RP_SIZE>;
   using Badge = std::vector<u8>;
+  static constexpr u32 BADGE_HW = 96;
 
   struct BadgeStatus
   {
     std::string name = "";
     Badge badge{};
+    std::shared_ptr<OSD::Icon> osd_icon;
```
Comment (AdmiralCurtiss) [inline Source/Core/Core/AchievementManager.h:72]: Is there a noticeable performance advantage to decoding these ahead of time rather than on-demand or what are you trying to accomplish with holding on to a second copy of what is effectively already stored in `badge`?

File: Source/Core/Core/AchievementManager.cpp [hunk 9]
```diff
@@ -11,18 +11,22 @@
...
 #include "Common/Logging/Log.h"
 #include "Common/WorkQueueThread.h"
 #include "Core/Config/AchievementSettings.h"
 #include "Core/Core.h"
 #include "Core/PowerPC/MMU.h"
 #include "Core/System.h"
 #include "DiscIO/Volume.h"
-#include "VideoCommon/OnScreenDisplay.h"
+#include "VideoCommon/AbstractGfx.h"
+#include "VideoCommon/Assets/CustomTextureData.h"
```
Comment (AdmiralCurtiss) [inline Source/Core/Core/AchievementManager.cpp:23]: These are not used.

File: Source/Core/Core/AchievementManager.cpp [hunk 10]
```diff
@@ -1184,6 +1209,30 @@ AchievementManager::PingRichPresence(const RichPresence& rich_presence)
...
+  {
+    OSD::AddMessage(fmt::format("You have {}/{} achievements worth {}/{} points",
+                                spread.hard_unlocks + spread.soft_unlocks, spread.total_count,
+                                spread.hard_points + spread.soft_points, spread.total_points),
+                    OSD::Duration::VERY_LONG, OSD::Color::CYAN,
+                    (Config::Get(Config::RA_BADGES_ENABLED)) ? m_game_badge.osd_icon : nullptr);
+    OSD::AddMessage("Hardcore mode is OFF", OSD::Duration::VERY_LONG, OSD::Color::CYAN);
+  }
+}
+
```
Comment (AdmiralCurtiss) [inline Source/Core/Core/AchievementManager.cpp:1235]: You didn't remove the equivalent code from `LoadGameByFilenameAsync()` that I assume you were intending to replace with this.

**Review by AdmiralCurtiss — COMMENTED** [review]
I question some of your design choices here.

File: Source/Core/VideoCommon/OnScreenDisplay.cpp (hunk 2 above)
Comment (LillyJadeKatrin) [reply Source/Core/VideoCommon/OnScreenDisplay.cpp:54]: It's a pointer I use in multiple places, I can't copy the unique_ptr and I can't move it because I might need it again, what am I misunderstanding about how this works?

Comment (LillyJadeKatrin) [reply Source/Core/VideoCommon/OnScreenDisplay.cpp:54]: Never mind, your other comment about why I'm not decoding when I need it resolves this.

File: Source/Core/Core/AchievementManager.cpp (hunk 10 above)
Comment (LillyJadeKatrin) [reply Source/Core/Core/AchievementManager.cpp:1235]: A lot of these were things from previous attempts I forgot to remove but this makes it obvious I borked a rebase somehow, my mistake.

**Comment by AdmiralCurtiss** [comment]
Your first commit didn't compile, I went ahead and fixed it.

The approach with the welcome message in the second commit is pretty janky, but I understand how you got there and I don't really have a better solution either -- most games just don't render anything for a few seconds on boot while they initialize, and Dolphin doesn't render anything when the game isn't. We should probably revisit this when and if async presentation exists.

**Comment by LillyJadeKatrin** [comment]
Thanks for the assist! Thought for sure I had nailed that down.
